cartographer/io/internal/mapping_state_serialization.cc

```diff
@@ -60,9 +60,86 @@ mapping::proto::SerializationHeader CreateHeader() {
   return header;
 }
 
-SerializedData SerializePoseGraph(const mapping::PoseGraph& pose_graph) {
+SerializedData SerializePoseGraph(const mapping::PoseGraph& pose_graph,
+                                  bool include_unfinished_submaps) {
   SerializedData proto;
   *proto.mutable_pose_graph() = pose_graph.ToProto();
+
+  if (include_unfinished_submaps) {
+    return proto;
+  }
+
+  std::set<mapping::SubmapId> unfinished_submaps;
+  std::set<mapping::NodeId> orphaned_nodes;
+
+  // Determine set of unfinished submaps.
+  for (const auto& submap_id_data : pose_graph.GetAllSubmapData()) {
+    if (!submap_id_data.data.submap->finished()) {
+      unfinished_submaps.insert(submap_id_data.id);
+    }
+  }
+
+  // Skip all those constraints in the proto that refer to unfinished
+  // submaps and remember the corresponding trajectory nodes as
+  // potentially orphaned.
+  std::vector<mapping::proto::PoseGraph::Constraint> constraints{
+      proto.pose_graph().constraint().begin(),
+      proto.pose_graph().constraint().end()};
+  proto.mutable_pose_graph()->clear_constraint();
+  for (auto it = constraints.begin(); it != constraints.end();) {
+    mapping::SubmapId submap_id(it->submap_id().trajectory_id(),
+                                it->submap_id().submap_index());
+    if (unfinished_submaps.count(submap_id) == 0) {
+      // Submap is finished.
+      *proto.mutable_pose_graph()->mutable_constraint()->Add() = *it;
+      ++it;
+    } else {
+      // Submap is unfinished.
+      mapping::NodeId node_id(it->node_id().trajectory_id(),
+                              it->node_id().node_index());
+      orphaned_nodes.insert(node_id);
+      it = constraints.erase(it);
+    }
+  }
+
+  // Iterate over all constraints and remove trajectory nodes from
+  // 'orphaned_nodes' that are not actually orphaned.
+  for (const auto& constraint : constraints) {
+    mapping::NodeId node_id(constraint.node_id().trajectory_id(),
+                            constraint.node_id().node_index());
+    orphaned_nodes.erase(node_id);
+  }
+
+  // Skip orphaned trajectory nodes and unfinished submaps in the
+  // trajectory proto.
+  for (mapping::proto::Trajectory& trajectory :
+       *proto.mutable_pose_graph()->mutable_trajectory()) {
+    // Skip unfinished submaps.
+    std::vector<mapping::proto::Trajectory::Submap> submaps{
+        trajectory.submap().begin(), trajectory.submap().end()};
+    trajectory.clear_submap();
+    for (const auto& submap : submaps) {
+      mapping::SubmapId submap_id(trajectory.trajectory_id(),
+                                  submap.submap_index());
+      if (unfinished_submaps.count(submap_id) == 0) {
+        // Submap is finished.
+        *trajectory.add_submap() = submap;
+      }
+    }
+
+    // Skip orphaned nodes.
```

Comment:
Add a test that covers this logic. It could be similar to PoseGraph3DTestBasicSerialization, or a map_builder_test that saves and loads some fake data. map_builder_test also has the last 1-2 submaps unfinished so that would be covered.